source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,693 @@
+      } else {
+        assert(false &&
+               "OperateVectors() only accepts VectorConst or NullConst type of "
+               "constant");
+      }
+    }
+    result.push_back(OperateWords(opcode, operand_values_for_one_dimension));
+  }
+  return result;
+}
+}  // anonymous namespace
+
+bool FoldSpecConstantOpAndCompositePass::ProcessImpl(ir::Module* module) {
+  bool modified = false;
+  auto inst_iter = module->types_values_begin();
+  // In case the types, constants, global variabls secion is empty, do nothing.
+  if (inst_iter == module->types_values_end()) return modified;
+
+  do {
+    // Collect constant values of normal constants and process the
+    // OpSpecConstantOp and OpSpecConstantComposite instructions if possible.
+    // OpConstantSampler instruction is not collected here because it cannot
+    // be used in OpSpecConstant{Composite|Op} instructions.
+    ir::Instruction* inst = &*inst_iter;
+    switch (SpvOp opcode = inst->opcode()) {
+      case SpvOp::SpvOpConstantTrue:
+      case SpvOp::SpvOpConstantFalse:
+      case SpvOp::SpvOpConstant:
+      case SpvOp::SpvOpConstantNull:
+      case SpvOp::SpvOpConstantComposite:
+      case SpvOp::SpvOpSpecConstantComposite: {
+        if (auto const_value = CreateConstFromInst(inst)) {
+          // Need to replace the OpSpecConstantComposite instruction with a
+          // corresponding OpConstantComposite instruction.
+          if (opcode == SpvOp::SpvOpSpecConstantComposite) {
+            inst->SetOpcode(SpvOp::SpvOpConstantComposite);
+            modified = true;
+          }
+          const_val_to_id_[const_value.get()] = inst->result_id();
+          id_to_const_val_[inst->result_id()] = std::move(const_value);
+        }
+        break;
+      }
+      case SpvOp::SpvOpSpecConstantOp:
+        modified |= ProcessOpSpecConstantOp(&inst_iter);
+        break;
+      default:
+        break;
+    }
+  } while (++inst_iter != module->types_values_end());
+  return modified;
+}
+
+bool FoldSpecConstantOpAndCompositePass::ProcessOpSpecConstantOp(
+    ir::Module::inst_iterator* pos) {
+  ir::Instruction* inst = &**pos;
+  ir::Instruction* folded_inst = nullptr;
+  assert(inst->GetInOperand(0).type ==
+             SPV_OPERAND_TYPE_SPEC_CONSTANT_OP_NUMBER &&
+         "The first in-operand of OpSpecContantOp instruction must be of "
+         "SPV_OPERAND_TYPE_SPEC_CONSTANT_OP_NUMBER type");
+
+  switch (static_cast<SpvOp>(inst->GetSingleWordInOperand(0))) {
+    case SpvOp::SpvOpCompositeExtract:
+      folded_inst = DoCompositeExtract(pos);
+      break;
+    case SpvOp::SpvOpVectorShuffle:
+      folded_inst = DoVectorShuffle(pos);
+      break;
+
+    case SpvOp::SpvOpCompositeInsert:
+      // Current Glslang does not generate code with OpSpecConstantOp
+      // CompositeInsert instruction, so this is not implmented so far.
+      // TODO(qining): Implement CompositeInsert case.
+      assert(false &&
```

Comment:
You raised this offline.  This should not assert, but just silently avoid trying the optimization.